### DISCUSSION TITLE
Better syntax for immutable types / C# 9 records

### DISCUSSION THREAD

**Discussion by jzabroski** [AutoFixture/AutoFixture#1212] [Ideas]
Currently, AutoFixture does not provide nice support for fluently creating C# 9 records (or any immutable objects).  In fact, with C# 9 records were announced, I thought, "Oh crap, a lot of my code won't be able to use AutoFixture if I use C# 9 records."

After researching this more thoroughly, it appears I was slightly wrong, here is the syntax that works:

```c#
// SUTs
public record FooRecord
{
	public int MyProperty { get; set; }
	public FooRecord(int myProperty) => (MyProperty) = (myProperty);
}

public class FooClass
{
	public int MyProperty { get; private set; }
	public FooClass(int myProperty)
	{
		MyProperty = myProperty;
	}
}

// Tests
var fixture = new AutoFixture.Fixture();
var fooClassCreate = fixture.Create<FooClass>();
var fooRecordCreate = fixture.Create<FooRecord>();
```

and, thus, you can do the following:

```c#
var fooWith = fixture.Create<FooRecord>() with { MyProperty = int.MaxValue };
```

However, by default, the following does not work:

```c#
// Tests
var fixture = new AutoFixture.Fixture();
fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
	.ForEach(b => fixture.Behaviors.Remove(b));
fixture.Behaviors.Add(new OmitOnRecursionBehavior());

var buildFooClassFromFactory = fixture
	.Build<FooClass>()
	.FromFactory((Func<FooClass> f) => new FooClass(fixture.Create<int>())).Create();
/* throws the following exception: 
AutoFixture was unable to create an instance from System.Func`1[FooClass] because creation unexpectedly failed with exception. Please refer to the inner exception to investigate the root cause of the failure.

Request path:
	System.Func`1[FooClass]

Inner Exception: Unable to cast object of type 'AutoFixture.Kernel.OmitSpecimen' to type 'System.Func`1[FooClass]'.
*/
```

This scenario is useful if you want to vary complex record object graphs.

**Comment by emilwall** [comment]
@jzabroski I think this is a great idea. I think support for [init only setters](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9#init-only-setters) in the Build API (see #1201)  would also be a great addition that perhaps ties in with your proposal.

**Reply by aivascu** [reply]
@emilwall not sure the solution for init-only member would apply well to record types. I mean it would work, you would probably get instances generated, but AutoFixture would be generating the values for the members twice.
The first generation would occur for the constructor, since record types do not have a parameterless constructor, and then the second generation of values would occur again for each member.
Not only this will potentially bring performance issue but also potential inconsistency in the generated values.

**Reply by emilwall** [reply]
Thank you for your input. I think I see what you mean, but in the example above with FooRecord the value is also generated twice so I don't really see the difference. Moreover, a constructor is not required for records when using init only setters. These are new things and I'm also still learning the details. Perhaps you are confused by how positional records work, since they use init only setters but also generate a constructor with arguments for the properties.
